src/legacy/core_plugins/kibana/public/dashboard/np_ready/dashboard_state_manager.ts

```diff
@@ -236,11 +245,20 @@ export class DashboardStateManager {
       ) {
         // A panel was changed
         dirty = true;
+
+        const oldVersion = savedDashboardPanelMap[panelState.explicitInput.id]?.version;
+        const newVersion = convertedPanelStateMap[panelState.explicitInput.id]?.version;
+        if (oldVersion !== newVersion) {
+          dirtyBecauseOfInitialStateMigration = true;
+        }
       }
     });
 
     if (dirty) {
       this.stateContainer.transitions.set('panels', Object.values(convertedPanelStateMap));
+      if (dirtyBecauseOfInitialStateMigration) {
```

Comment:
Not sure whether I misunderstand, but it this code path is also triggered when a new panel is added - this has the effect that the history entry of the state change adding the panel is overwriting the previous state change.

Example:
* Go to dashboard
* Enter edit mode
* Add 2 panels
* Click back
* Dashboard removes both panels and leaves edit mode

Seems like the check above has to be 
```ts
if (oldVersion && newVersion && oldVersion !== newVersion) {
```

But I haven't checked

Reply:
Seems you are right, nice catch